### PR TITLE
* Change `IConnection` event `ConnectionShutdown` to `AsyncEventHandler`

### PR DESCRIPTION
**Pull request by lukebakken**
* Remove sync methods from `IConsumerDispatcher`
* Make several `ISession` methods async

### CODE REVIEW

File: projects/RabbitMQ.Client/client/impl/ChannelBase.cs [hunk 1]
```diff
@@ -506,14 +508,12 @@ private void OnChannelShutdown(ShutdownEventArgs reason)
...
-        // TODO async
-        private void OnSessionShutdown(object sender, ShutdownEventArgs reason)
+        private Task OnSessionShutdownAsync(object sender, ShutdownEventArgs reason)
         {
             ConsumerDispatcher.Quiesce();
             SetCloseReason(reason);
             OnChannelShutdown(reason);
-            // TODO async
-            ConsumerDispatcher.Shutdown(reason);
+            return ConsumerDispatcher.ShutdownAsync(reason);
```
Comment (JanEggers) [inline projects/RabbitMQ.Client/client/impl/ChannelBase.cs:516]: missiing .ConfigureAwait(false);

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs [hunk 2]
```diff
@@ -175,26 +178,29 @@ private void MaybeTerminateMainloopAndStopHeartbeatTimers(bool cancelMainLoop =
...
-                return;
+                return Task.CompletedTask;
             }
 
             _channel0.MaybeSetConnectionStartException(reason.Exception);
 
-            OnShutdown(reason);
             LogCloseError($"Unexpected connection closure: {message}", reason.Exception);
+
+            return OnShutdownAsync(reason);
```
Comment (JanEggers) [inline projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:194]: missing .ConfigureAwait(false);

File: projects/RabbitMQ.Client/client/impl/ChannelBase.cs (hunk 1 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/ChannelBase.cs:515]: Thanks, I have no idea why the `CA2007` check didn't catch this.

Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/ChannelBase.cs:515]: Oh, I'm not actually awaiting the `Task` here, just returning it.

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs (hunk 2 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:194]: Same as above ... the `Task` is not awaited here.

**Pull request closed without merging**